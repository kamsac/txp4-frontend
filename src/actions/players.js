import dotenvConfiguration from '../dotenv-configuration';
import PlayerResourceMock from '../resources/players/mock';
import PlayerResource from '../resources/players';

export const PLAYER_SUCCESS = 'PLAYER_SUCCESS';
export const PLAYER_REQUEST = 'PLAYER_REQUEST';
export const PLAYER_FAILURE = 'PLAYER_FAILURE';

export function requestPlayer() {
  return {
    type: PLAYER_REQUEST,
    payload: {
      isFetching: true,
    },
  };
}

export function receivePlayer(player) {
  return {
    type: PLAYER_SUCCESS,
    payload: {
      isFetching: false,
      player,
    },
  };
}

export function playerError() {
  return {
    type: PLAYER_FAILURE,
    payload: {
      isFetching: false,
    },
  };
}

export function loadPlayer(playerLogin) {
  return (dispatch) => {
    dispatch(requestPlayer());

    const resource = dotenvConfiguration.API_URL ? PlayerResource : PlayerResourceMock;

    resource.getPlayer(playerLogin)
      .then((response) => {
        dispatch(receivePlayer(response.data.player));
      })
      .catch(() => {
        dispatch(playerError());
      });
  };
}
