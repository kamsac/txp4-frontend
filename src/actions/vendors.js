import VendorsResource from '../resources/vendors/mock';

export const REQUEST_VENDORS = 'vendors:requestVendors';
export function requestVendors() {
  return dispatch => {
    return dispatch(fetchVendors());
  }
}

function fetchVendors() {
  return dispatch => {
    VendorsResource.getVendors((vendors) => {
      dispatch(receiveVendors(vendors));
    });
  }
}

export const RECEIVE_VENDORS = 'vendors:fetchVendorsSuccess';
export function receiveVendors(vendors) {
  return {
    type: RECEIVE_VENDORS,
    vendors: vendors,
  }
}
