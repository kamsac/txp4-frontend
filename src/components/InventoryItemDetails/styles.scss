@import '../../variables';

.InventoryItemDetails {
  border: $border-width solid $black;
  border-radius: $border-radius;
  grid-area: InventoryItemDetails;
}
