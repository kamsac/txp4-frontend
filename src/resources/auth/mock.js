export const ACCESS_TOKEN = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiaWF0IjoxNTE2MjM5MDIyLCJsb2dpbiI6Imx5bWFrIiwibmljayI6IiRzJG8kYTYwbCRjODJ5JGZhM20kYzgyYSRhNjBrIn0.aaKzXEZP2tovhBAqPJYBdmbLzKjfdBsuX3FaB58JZvE';

const AuthResourceMock = {
  login() {
    return new Promise((resolve) => {
      const response = {
        data: {
          token: ACCESS_TOKEN,
        },
      };
      setTimeout(() => resolve(response), 100);
    });
  },
};

export default AuthResourceMock;
