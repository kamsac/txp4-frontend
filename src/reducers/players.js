import inventoryReducer from './inventory';
import { PLAYER_FAILURE, PLAYER_REQUEST, PLAYER_SUCCESS } from '../actions/players';
import {
  INVENTORY_REQUEST,
  INVENTORY_FAILURE,
  INVENTORY_SUCCESS,
  EQUIP_REQUEST,
  EQUIP_SUCCESS,
  EQUIP_FAILURE,
} from '../actions/inventory';

const initialState = {
  isFetching: false,
  items: {},
};

export default function playersReducer(state = initialState, action) {
  switch (action.type) {
    case PLAYER_REQUEST: {
      return Object.assign({}, state, {
        isFetching: action.payload.isFetching,
      });
    }
    case PLAYER_SUCCESS: {
      const { isFetching, player } = action.payload;
      return Object.assign({}, state, {
        isFetching,
        items: Object.assign({}, state.items, {
          [player.login]: {
            login: player.login,
            nick: player.nick,
            inventory: {
              isFetching: false,
              value: player.inventory,
            },
          },
        }),
      });
    }
    case PLAYER_FAILURE: {
      const { isFetching } = action.payload;
      return Object.assign({}, state, {
        isFetching,
      });
    }

    case INVENTORY_REQUEST:
    case INVENTORY_SUCCESS:
    case INVENTORY_FAILURE:
    case EQUIP_REQUEST:
    case EQUIP_SUCCESS:
    case EQUIP_FAILURE: {
      const { playerLogin } = action.payload;

      const nextState = Object.assign({}, state);

      if (!nextState.items[playerLogin]) {
        nextState.items[playerLogin] = {};
      }

      const inventory = inventoryReducer(nextState.items[playerLogin].inventory, action);
      nextState.items[playerLogin].inventory = inventory;
      return nextState;
    }
    default:
      return state;
  }
}

export function getPlayer(state, playerLogin) {
  if (
    state.players &&
    state.players.items
  ) {
    return state.players.items[playerLogin];
  }

  return null;
}

export function getPlayerNick(state, playerLogin) {
  if (
    state.players &&
    state.players.items &&
    state.players.items[playerLogin]
  ) {
    return state.players.items[playerLogin].nick;
  }

  return null;
}
