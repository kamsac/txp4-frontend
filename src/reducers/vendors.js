import {RECEIVE_VENDORS} from '../actions/vendors';

const initialValue = [];

export function vendorsReducer(state = initialValue, action) {
  switch(action.type) {
    case RECEIVE_VENDORS:
      return [...action.vendors];
    default:
      return state;
  }
}
