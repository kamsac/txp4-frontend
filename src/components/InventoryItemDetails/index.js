import React from 'react';
import './styles.scss';
import InventoryService from '../../resources/inventory/mock';
import InventoryItemModifiers from '../InventoryItemModifiers';
import InventoryItemTitle from '../InventoryItemTitle';

export default class InventoryItemDetails extends React.Component {
  constructor(props) {
    super(props);
  }

  handleEquipItem(item, event) {
    InventoryService.equipItem(item, (response) => {
      this.props.onEquipItem(item);
    });
  }

  render() {
    const item = this.props.item;

    if (item) {

      return (
        <div className="InventoryItemDetails">
          <InventoryItemTitle item={item}/>
          <InventoryItemModifiers item={item}/>

          <button onClick={(event) => this.handleEquipItem(item, event)}>Equip this item</button>
        </div>
      );
    } else {
      return (
        <div className="InventoryItemDetails">
          <p>Select item to display details.</p>
        </div>
      )
    }
  }
}
