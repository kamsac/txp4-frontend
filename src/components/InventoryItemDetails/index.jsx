import React from 'react';
import PropTypes from 'prop-types';
import './styles.scss';
import InventoryItemModifiers from '../InventoryItemModifiers/index';
import InventoryItemTitle from '../InventoryItemTitle/index';
import { ItemPropTypesShape, VendorPropTypesShape } from '../../prop-types';
import Button from '../Button';

export default class InventoryItemDetails extends React.Component {
  render() {
    const { item, showControls } = this.props;

    if (item) {
      return (
        <div className="InventoryItemDetails">
          <InventoryItemTitle item={item} vendors={this.props.vendors} />
          <InventoryItemModifiers item={item} />
          {showControls &&
            <Button type="primary" onClick={() => this.props.onEquipItem(item)}>Equip this item</Button>
          }
        </div>
      );
    }
    return (
      <div className="InventoryItemDetails">
        <p>Select item to display details.</p>
      </div>
    );
  }
}

InventoryItemDetails.propTypes = {
  item: ItemPropTypesShape,
  vendors: PropTypes.arrayOf(VendorPropTypesShape).isRequired,
  onEquipItem: PropTypes.func.isRequired,
  showControls: PropTypes.bool.isRequired,
};

InventoryItemDetails.defaultProps = {
  item: null,
};
