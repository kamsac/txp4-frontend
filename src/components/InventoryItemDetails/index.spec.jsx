import React from 'react';
import { shallow } from 'enzyme';
import InventoryItemDetails from './index';
import { ITEMS } from '../../resources/players/mock';
import { VENDORS } from '../../resources/vendors/mock';

describe('<InventoryItemDetails>', () => {
  test('with selected item', () => {
    const component = shallow(<InventoryItemDetails
      item={ITEMS[0]}
      vendors={VENDORS}
      onEquipItem={jest.fn()}
      showControls
    />);
    expect(component).toMatchSnapshot();
  });

  test('with no selected item', () => {
    const component = shallow(<InventoryItemDetails
      item={null}
      vendors={VENDORS}
      onEquipItem={jest.fn()}
      showControls
    />);
    expect(component).toMatchSnapshot();
  });

  it('should equip item', () => {
    const onEquipItem = jest.fn();
    const component = shallow(<InventoryItemDetails
      item={ITEMS[0]}
      vendors={VENDORS}
      onEquipItem={onEquipItem}
      showControls
    />);
    const equipItemButton = component.find('button').filterWhere(wrapper => wrapper.text() === 'Equip this item');
    equipItemButton.simulate('click');
    expect(onEquipItem).toHaveBeenCalled();
  });

  it('should not render equip button when user is not owner', () => {
    const component = shallow(<InventoryItemDetails
      item={ITEMS[0]}
      vendors={VENDORS}
      onEquipItem={jest.fn()}
      showControls={false}
    />);
    expect(component).toMatchSnapshot();
  });
});
