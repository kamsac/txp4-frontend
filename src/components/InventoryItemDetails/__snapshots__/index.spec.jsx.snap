// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InventoryItemDetails> should not render equip button when user is not owner 1`] = `
<div
  className="InventoryItemDetails"
>
  <InventoryItemTitle
    item={
      Object {
        "equipRegion": "tire",
        "id": "a1a1a1",
        "isEquipped": true,
        "modifiers": Array [
          Object {
            "name": "maxSpeed",
            "value": 1,
          },
          Object {
            "name": "acceleration",
            "value": 1.2,
          },
          Object {
            "name": "steering",
            "value": 1.1,
          },
          Object {
            "name": "breaking",
            "value": 1.2,
          },
          Object {
            "name": "grip",
            "value": 1.5,
          },
          Object {
            "name": "gravity",
            "value": 1,
          },
        ],
        "name": "Tire",
        "tier": 3,
        "vendorId": "aaa",
      }
    }
    vendors={
      Array [
        Object {
          "color": "#8f0",
          "id": "aaa",
          "name": "Super",
        },
        Object {
          "color": "#f80",
          "id": "bbb",
          "name": "Safe",
        },
        Object {
          "color": "#800",
          "id": "ccc",
          "name": "Heavy",
        },
        Object {
          "color": "#08f",
          "id": "ddd",
          "name": "Light",
        },
      ]
    }
  />
  <InventoryItemModifiers
    item={
      Object {
        "equipRegion": "tire",
        "id": "a1a1a1",
        "isEquipped": true,
        "modifiers": Array [
          Object {
            "name": "maxSpeed",
            "value": 1,
          },
          Object {
            "name": "acceleration",
            "value": 1.2,
          },
          Object {
            "name": "steering",
            "value": 1.1,
          },
          Object {
            "name": "breaking",
            "value": 1.2,
          },
          Object {
            "name": "grip",
            "value": 1.5,
          },
          Object {
            "name": "gravity",
            "value": 1,
          },
        ],
        "name": "Tire",
        "tier": 3,
        "vendorId": "aaa",
      }
    }
  />
</div>
`;

exports[`<InventoryItemDetails> with no selected item 1`] = `
<div
  className="InventoryItemDetails"
>
  <p>
    Select item to display details.
  </p>
</div>
`;

exports[`<InventoryItemDetails> with selected item 1`] = `
<div
  className="InventoryItemDetails"
>
  <InventoryItemTitle
    item={
      Object {
        "equipRegion": "tire",
        "id": "a1a1a1",
        "isEquipped": true,
        "modifiers": Array [
          Object {
            "name": "maxSpeed",
            "value": 1,
          },
          Object {
            "name": "acceleration",
            "value": 1.2,
          },
          Object {
            "name": "steering",
            "value": 1.1,
          },
          Object {
            "name": "breaking",
            "value": 1.2,
          },
          Object {
            "name": "grip",
            "value": 1.5,
          },
          Object {
            "name": "gravity",
            "value": 1,
          },
        ],
        "name": "Tire",
        "tier": 3,
        "vendorId": "aaa",
      }
    }
    vendors={
      Array [
        Object {
          "color": "#8f0",
          "id": "aaa",
          "name": "Super",
        },
        Object {
          "color": "#f80",
          "id": "bbb",
          "name": "Safe",
        },
        Object {
          "color": "#800",
          "id": "ccc",
          "name": "Heavy",
        },
        Object {
          "color": "#08f",
          "id": "ddd",
          "name": "Light",
        },
      ]
    }
  />
  <InventoryItemModifiers
    item={
      Object {
        "equipRegion": "tire",
        "id": "a1a1a1",
        "isEquipped": true,
        "modifiers": Array [
          Object {
            "name": "maxSpeed",
            "value": 1,
          },
          Object {
            "name": "acceleration",
            "value": 1.2,
          },
          Object {
            "name": "steering",
            "value": 1.1,
          },
          Object {
            "name": "breaking",
            "value": 1.2,
          },
          Object {
            "name": "grip",
            "value": 1.5,
          },
          Object {
            "name": "gravity",
            "value": 1,
          },
        ],
        "name": "Tire",
        "tier": 3,
        "vendorId": "aaa",
      }
    }
  />
  <button
    onClick={[Function]}
  >
    Equip this item
  </button>
</div>
`;
